src/dev/lupio/lupio_tty.cc

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2021 The Regents of the University of California
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dev/lupio/lupio_tty.hh"
+
+#include "base/trace.hh"
+#include "debug/LupioTTY.hh"
+#include "dev/platform.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioTTY.hh"
+
+#define LUPIO_TTY_INVAL    0x80000000
+
+/* Same fields for CTRL and STAT registers */
+#define LUPIO_TTY_WBIT  (1 << 0)
+#define LUPIO_TTY_RBIT  (1 << 1)
+
+namespace gem5
+{
+
+LupioTTY::LupioTTY(const Params &params)
+    : BasicPioDevice(params, params.pio_size),
+    writChar(-1), readChar(-1), writIntrEn(false), readIntrEn(false),
+    device(params.device), platform(params.platform)
+{
+    // setup serial device callbacks
+    device->regInterfaceCallback([this]() { dataAvailable(); });
+}
+
+void
+LupioTTY::lupioTTYUpdateIRQ()
+{
+    unsigned int irq;
+
+    irq = (writIntrEn && writChar != -1)
+        || (readIntrEn && readChar != -1);
+
+    if (irq) {
+        DPRINTF(LupioTTY, "LupioTTY InterEvent, interrupting\n");
+        platform->postConsoleInt();
+    } else {
+        DPRINTF(LupioTTY, "LupioTTY InterEvent, clear interrupt\n");
+        platform->clearConsoleInt();
+    }
+}
+
+void
+LupioTTY::dataAvailable()
+{
+    if (device->dataAvailable()) {
```

Comment:
I wonder if that should really be an `if`. If this function is called, it *should* be because there is data available. The lack of data would then be an error, caught as such:

```C
assert(device->dataAvailable());
```

Reply:
`chatty_assert` is preferred: https://gem5.googlesource.com/public/gem5/+/refs/heads/develop/src/base/logging.hh#304

Reply:
Actually, now it's just `gem5_assert` if you rebase on current develop: https://gem5-review.googlesource.com/c/public/gem5/+/48606